### PR TITLE
Adaptive filter reordering in scan

### PR DESCRIPTION
**Pull request by mbasmanova**
We are seeing a performance regression for filters with multiple conjuncts on different columns and 
the first conjunct being highly selective.

`SELECT * FROM t WHERE f(a) AND g(b)`

In these cases, in baseline, LazyBlocks for the 2nd conjunct are not loaded (or loaded rarely). In Aria, we had all filter functions evaluated together after all inputs for all of them are read. In the query above, we read columns a and b, then evaluate f(a), then f(b). 

This PR introduces a few changes:
- Evaluate each filter function as soon as all the inputs are read. This way, f(a) above will be evaluated before column b is read.
- Change the initial order of columns to (1) columns with range filters; (2) columns providing inputs to filter functions; (3) columns without filters. In group (2), columns are ordered by type (simply types first): bool, integers, floats, strings, arrays/maps/structs.
- Introduce adaptive filter reordering. Track number of rows dropped per unit of time for each function, then put most productive (drops rows fastest) first.

CC: @prestodb/aria 

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java [hunk 1]
```diff
@@ -817,4 +823,55 @@ public final void load(LazyBlock lazyBlock)
...
+        {
+            return stats;
+        }
+    }
+
+    private static final class FilterStats
+    {
+        private long inputPositions;
+        private long outputPositions;
+        private long time;
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java]: We usually use names like `elapsedNanos`, `elapsedMillis`, etc.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java [hunk 2]
```diff
@@ -817,4 +823,55 @@ public final void load(LazyBlock lazyBlock)
...
+
+        public void update(int inputPositions, int outputPositions, long time)
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java]: how about `increase`?

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java [hunk 3]
```diff
@@ -229,19 +246,127 @@ public OrcSelectiveRecordReader(
...
+            Object constantValue = constantValues[columnIndex];
+            blocks[i] = RunLengthEncodedBlock.create(columnTypes.get(columnIndex), constantValue == NULL_MARKER ? null : constantValue, 1);
+        }
+
+        int[] positions = new int[] {0};
+        initializeTmpErrors(1);
+        int positionCount = function.filter(new Page(blocks), positions, 1, tmpErrors);
+
+        if (tmpErrors[0] != null) {
+            this.constantFilterError = tmpErrors[0];
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java]: remove "this."

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java [hunk 4]
```diff
@@ -229,19 +246,127 @@ public OrcSelectiveRecordReader(
...
+            }
+            order[lastIndex].add(function);
+        }
+
+        return order;
+    }
+
+    private static Set<Integer>[] collectFilterFunctionInputs(List<FilterFunction>[] functionsOrder, Map<Integer, Integer> inputMapping)
+    {
+        Set[] inputs = new Set[functionsOrder.length];
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java]: `Set<Integer>[] inputs = new Set[functionsOrder.length];`

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java [hunk 5]
```diff
@@ -229,19 +246,127 @@ public OrcSelectiveRecordReader(
...
+                .flatMapToInt(function -> Arrays.stream(function.getInputChannels()))
+                .boxed()
+                .map(this.filterFunctionInputMapping::get)
+                .collect(toImmutableSet());
+    }
+
+    private boolean evaluateDeterministicFilterFunctionsWithConstantInputs(List<FilterFunction> filterFunctions)
+    {
+        for (FilterFunction function : filterFunctions) {
+            if (function.isDeterministic() && hasInputs(function) && allConstantInputs(function)) {
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java]:
```java
if (function.isDeterministic() && hasInputs(function) && allConstantInputs(function) && !evaluateDeterministicFilterFunctionWithConstantInputs(function)) {
    return false;
}

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java [hunk 6]
```diff
@@ -229,19 +246,127 @@ public OrcSelectiveRecordReader(
...
+    {
+        int[] inputs = function.getInputChannels();
+        Block[] blocks = new Block[inputs.length];
+        for (int i = 0; i < inputs.length; i++) {
+            int columnIndex = filterFunctionInputMapping.get(inputs[i]);
+            Object constantValue = constantValues[columnIndex];
+            blocks[i] = RunLengthEncodedBlock.create(columnTypes.get(columnIndex), constantValue == NULL_MARKER ? null : constantValue, 1);
+        }
+
+        int[] positions = new int[] {0};
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java]: this variable can be inlined

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java [hunk 7]
```diff
@@ -86,24 +90,40 @@
...
+    private Set<Integer> filterFunctionConstantInputs;
+
     // An immutable list of initial positions; includes all positions: 0,1,2,3,4,..
     // This array may grow, but cannot shrink. The values don't change.
     private int[] positions;
 
     // Used in applyFilterFunctions; mutable
     private int[] outputPositions;
     private RuntimeException[] errors;
+    private RuntimeException[] tmpErrors;
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java]: How about `intermediateErrors`? Maybe add a comment to explain the difference from the above

Comment (mbasmanova) [reply presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java]: @highker The difference is that `errors` contains shared state (used across method calls), but `tmpErrors` is just a local variable (I'm using member variable only to enable memory re-use). How about this?

```
    // errors encountered while evaluating filter functions; indices are positions in the batch
    // of rows being processed by getNextPage (errors[outputPositions[i]] is valid)
    private RuntimeException[] errors;

    // temporary array to be used in applyFilterFunctions only; exists solely for the purpose of re-using memory
    // indices are positions in a page provided to the filter filters (it contains a subset of rows that passed earlier filters)
    private RuntimeException[] tmpErrors;
```

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java [hunk 8]
```diff
@@ -216,19 +259,182 @@ public OrcSelectiveRecordReader(
...
+        // filterFunctionsOrder (aligned with streamReaderOrder),
+        // filterFunctionInputs (aligned with filterFunctionsOrder)
+        boolean sameOrder = true;
+        for (int i = 0; i < streamReaderOrder.length; i++) {
+            if (!columnsWithFilterScores.containsKey(streamReaderOrder[i])) {
+                for (int j = 0; j < newColumnOrder.length; j++) {
+                    if (streamReaderOrder[i] != newColumnOrder[j]) {
+                        sameOrder = false;
+                    }
+                    streamReaderOrder[i++] = newColumnOrder[j];
```
Comment (bhhari) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java]: is "i" getting double incremented in the loop.
Also why don't we inc both the arrays in a single loop
for(int i,j=0;condition;i++,j++)

Comment (mbasmanova) [reply presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java]: @bhhari `streamReaderOrder` contains 3 groups of columns, one after another: 

- columns with range filters;
- columns that provide inputs to filter functions;
- columns with no filter.

`newColumnOrder` contains columns from the second group in the "right" order.

The outer loop goes over streamReaderOrder until it reaches first column from the second group, then it loops over newColumnOrder to update columns in the second group, then exits.

**Pull request merged**